src/main/java/com/google/sps/servlets/BooksServlet.java

```diff
@@ -0,0 +1,142 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import com.google.gson.Gson;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.sps.data.Book;
+import java.net.URL;
+import java.net.HttpURLConnection;
+import java.util.Scanner;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonElement;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Servlet for generating books from keywords.
+ */
+@WebServlet("/books")
+public class BooksServlet extends HttpServlet {
+
+  private final int NUM_BOOKS_PER_KEYWORD = 5;
+  private LinkedHashMap<String, List<Book>> booksMap;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    List<String> keywords = Arrays.asList(request.getParameterValues("key"));
+    booksMap = new LinkedHashMap<>();
+    keywords.forEach((keyword) -> addBooksForTerm(keyword));
+    String json = encodeBookMapAsJson();
+    response.getWriter().println(json);
+  }
+
+ /**
+  * Encodes the given term so that it can fit in a url.
+  * @return An encoded string
+  */
+  private String encodeTerm(String term) {
+    try {
+      return URLEncoder.encode(term, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
+  private String encodeBookMapAsJson() {
+    Gson gson = new Gson();
+    Map<String, Object> books = new HashMap<>();
+    books.put("books", booksMap);
+    String json = gson.toJson(books); 
+    return json;
+  }
+
+ /**
+  * Adds Book object to books from JSON results of Google Books query.
+  */
+  private void addBooksToMap(String jsonString, String keyTerm) {
+    JsonObject responseObject = new JsonParser().parse(jsonString).getAsJsonObject();
+    // Check to see if json has enough entries.
+    int numResults = responseObject.get("totalItems").getAsInt();
+    numResults = Math.min(numResults, NUM_BOOKS_PER_KEYWORD);
+    if (numResults <= 0) {
```

Comment:
Do you need to do this. I assume there will be many terms that produce no results so you need to gracefully deal with that. If you just removed this whole if block the function would put <'term, [ empty ] > into bookMap and then whatever code has to present that information to the user can figure out how to tell the user that term had no results.

What do you think?

Reply:
I think you're right -- the Mustache template handles it well when there are no results, so there's not really a reason to print an error if there are no results.